### PR TITLE
added boilerplate for user management

### PR DESCRIPTION
**Pull request by Vasilije1990**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new job for setting up Docker Buildx in CI/CD workflows for efficient multi-platform builds.
	- Added a PostgreSQL service to the CI/CD workflow, improving the testing environment for database interactions.
	- Implemented a new step to ensure PostgreSQL is fully initialized before running tests, enhancing reliability.

- **Bug Fixes**
	- Improved error handling and logging for user-related actions.

- **Documentation**
	- Updated user management API documentation to reflect new endpoints and functionalities.

### CODE REVIEW

File: cognee/api/client.py [hunk 1]
```diff
@@ -267,8 +345,8 @@ def start_api_server(host: str = "0.0.0.0", port: int = 8000):
         relational_config.create_engine()
 
         from cognee.modules.data.deletion import prune_system, prune_data
-        asyncio.run(prune_data())
-        asyncio.run(prune_system(metadata = True))
+        # asyncio.run(prune_data())
```
Comment (borisarzentar) [inline cognee/api/client.py:348]: Please uncomment this after you are done.

File: cognee/api/v1/cognify/cognify_v2.py [hunk 2]
```diff
@@ -25,7 +31,21 @@
...
 update_status_lock = asyncio.Lock()
 
-async def cognify(datasets: Union[str, list[str]] = None, root_node_id: str = None):
+class PermissionDeniedException(Exception):
+    def __init__(self, message: str):
+        self.message = message
+        super().__init__(self.message)
+
+async def cognify(datasets: Union[str, list[str]] = None, root_node_id: str = None, user_id:str="default_user"):
+    session: AsyncSession = async_session_maker()
```
Comment (borisarzentar) [inline cognee/api/v1/cognify/cognify_v2.py:40]: Let's put this login in a function in some module and just use it here.
Something like `def check_permissions(user)` for example.

File: cognee/infrastructure/databases/relational/sqlalchemy/SqlAlchemyAdapter.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+class SQLAlchemyAdapter():
+    def __init__(self, db_type: str, db_path: str, db_name: str, db_user:str, db_password:str, db_host:str, db_port:str):
+        self.db_location = os.path.abspath(os.path.join(db_path, db_name))
+        # self.engine = create_engine(f"{db_type}:///{self.db_location}")
+        if db_type == "duckdb":
+            self.engine = create_engine(f"duckdb:///{self.db_location}")
+            self.Session = sessionmaker(bind=self.engine)
+
+        else:
+            print("Name: ", db_name)
```
Comment (borisarzentar) [inline cognee/infrastructure/databases/relational/sqlalchemy/SqlAlchemyAdapter.py:15]: Let's remove print statements

**Review by borisarzentar — CHANGES_REQUESTED** [review]

File: cognee/infrastructure/databases/relational/user_authentication/authentication_db.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+class Base(DeclarativeBase):
+    pass
+
+
+class User(SQLAlchemyBaseUserTableUUID, Base):
+    pass
+
+
+
+llm_config = get_relationaldb_config()
```
[the remark this answers is not in the document]
Comment (Vasilije1990) [reply cognee/infrastructure/databases/relational/user_authentication/authentication_db.py]: fixed this

File: cognee/infrastructure/databases/relational/user_authentication/authentication_db.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+# Association table for many-to-many relationship between groups and permissions
+group_permission = Table('group_permission', Base.metadata,
+                         Column('group_id', UUID, ForeignKey('groups.id')),
+                         Column('permission_id', UUID, ForeignKey('permissions.id')))
+
+
+
+class User(SQLAlchemyBaseUserTableUUID, Base):
+    __tablename__ = 'users'
+    groups = relationship('Group', secondary=user_group, back_populates='users')
```
[the remark this answers is not in the document]
Comment (Vasilije1990) [reply cognee/infrastructure/databases/relational/user_authentication/authentication_db.py:28]: it already does this  when it inherits from SQLAlchemyBaseUserTableUUID

File: cognee/api/v1/cognify/cognify_v2.py (hunk 2 above)
Comment (Vasilije1990) [reply cognee/api/v1/cognify/cognify_v2.py:40]: Let's discuss this live

File: cognee/infrastructure/databases/relational/user_authentication/authentication_db.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+from sqlalchemy.ext.asyncio import AsyncSession, async_sessionmaker, create_async_engine
+from sqlalchemy.orm import DeclarativeBase, sessionmaker, Session
+from cognee.infrastructure.databases.relational import get_relationaldb_config, Base
+from sqlalchemy import Column, String, ForeignKey, Table, Integer
+
+
+from sqlalchemy.orm import DeclarativeBase, sessionmaker, Session, relationship
+
+# Association table for many-to-many relationship between users and groups
+user_group = Table('user_group', Base.metadata,
```
[the remark this answers is not in the document]
Comment (Vasilije1990) [reply cognee/infrastructure/databases/relational/user_authentication/authentication_db.py:15]: Can you specify exactly what needs to be done?

File: cognee/infrastructure/databases/relational/user_authentication/routers.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+permission_router = APIRouter()
+
+@permission_router.post("/groups/{group_id}/permissions")
+async def assign_permission_to_group(group_id: int, permission: str, db: Session = Depends(get_user_db)):
+    group = db.query(Group).filter(Group.id == group_id).first()
+    if not group:
+        raise HTTPException(status_code=404, detail="Group not found")
+    perm = db.query(Permission).filter(Permission.name == permission).first()
+    if not perm:
```
[the remark this answers is not in the document]
Comment (Vasilije1990) [reply cognee/infrastructure/databases/relational/user_authentication/routers.py:13]: Don't get it, what is the exact problem?

File: cognee/infrastructure/databases/relational/user_authentication/users.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+current_active_user = fastapi_users.current_user(active=True)
+
+
+async def get_user_permissions(user: User, session: Session):
+    permissions = set()
+    for group in user.groups:
+        permissions.update(permission.name for permission in group.permissions)
+    return permissions
+
+def has_permission(permission: str):
```
[the remark this answers is not in the document]
Comment (Vasilije1990) [reply cognee/infrastructure/databases/relational/user_authentication/users.py:65]: Where exactly?

**Pull request merged**